csrc/index_compute.cpp

```diff
@@ -3171,4 +3171,83 @@ Val* Index::eye(
   return GpuLower::current()->commonScalarMap().hoistScalar(result, loops);
 }
 
+Val* Index::cpAsyncBulkIndex(
+    TensorView* tv,
+    const std::vector<kir::ForLoop*>& loops) {
+  using namespace tma;
+
+  TORCH_INTERNAL_ASSERT(
+      tv->getMemoryType() == MemoryType::Global,
+      "cpAsyncBulkIndex is only for global memory tensors");
+  TORCH_INTERNAL_ASSERT(
+      tv->getMaybeRFactorDomain() == tv->getLeafDomain(), "not supported yet");
+  TORCH_INTERNAL_ASSERT(
+      tv->getMaybeAllocationDomain() == tv->getLeafDomain(),
+      "not supported yet");
+  for (auto id : tv->getMaybeRFactorDomain()) {
+    TORCH_INTERNAL_ASSERT(
+        id->isBulk(),
+        "cpAsyncBulkIndex only support whole tensor copy for now.");
+  }
+
+  int64_t dim = (int64_t)tv->nDims();
+  TORCH_INTERNAL_ASSERT(dim > 0);
+  int64_t itemsize = dataTypeSize(tv->dtype());
+
+  auto metadata = IrBuilder::metadataExpr(tv);
+  auto global_address = IrBuilder::getAttrExpr(metadata, "data");
+  auto global_dim =
+      // Reverse array to convert from row major to column major
+      IrBuilder::reverseArrayExpr(
+          IrBuilder::getAttrExpr(metadata, "alloc_size"));
+  auto global_strides = IrBuilder::getAttrExpr(metadata, "alloc_stride");
+  if (dim > 1) {
+    // Reverse array to convert from row major to column major, multiply by
+    // element size to convert to bytes, and remove fastest dim as it is assumed
+    // to be one.
+    std::vector<Val*> strides;
+    for (auto i : c10::irange(dim - 1)) {
+      strides.push_back(SimplifyingIrBuilder::mulExpr(
+          IrBuilder::getItemExpr(global_strides, dim - 2 - i), itemsize));
```

Comment:
I think nvFuser supports constant negative axis, for example `tv->merge(-2)`, but not symbolic negative axis. For example, you can not have a `Val` that has runtime value `-1` and use that `Val` to index a tensor. For this case, because we are creating an expression, it is closer to the symbolic case, so I prefer to not use or support any negative axis.